homeassistant/components/rfxtrx/config_flow.py

```diff
@@ -1,12 +1,294 @@
+    DEVICE_PACKET_TYPE_LIGHTING4,
+)
+from .cover import supported as cover_supported
+from .light import supported as light_supported
+from .switch import supported as switch_supported
 
 _LOGGER = logging.getLogger(__name__)
 
+CONF_OFF_DELAY_ENABLED = CONF_OFF_DELAY + "_enabled"
+CONF_EVENT_CODE = "event_code"
+
+
+class OptionsFlow(config_entries.OptionsFlow):
+    """Handle Rfxtrx options."""
+
+    def __init__(self, config_entry: ConfigEntry) -> None:
+        """Initialize rfxtrx options flow."""
+        self._config_entry = config_entry
+        self._global_options = None
+        self._selected_device = None
+        self._selected_device_event_code = None
+        self._selected_device_object = None
+        self._device_entries = None
+
+    async def async_step_init(self, user_input=None):
+        """Manage the options."""
+        return await self.async_step_prompt_options()
+
+    async def async_step_prompt_options(self, user_input=None):
+        """Prompt for options."""
+        errors = {}
+
+        if user_input is not None:
+            self._global_options = {
+                CONF_DEBUG: user_input[CONF_DEBUG],
+                CONF_AUTOMATIC_ADD: user_input[CONF_AUTOMATIC_ADD],
+            }
+            if CONF_DEVICE in user_input:
+                device_data = self._get_device_data(user_input[CONF_DEVICE])
+                if not device_data.get(CONF_EVENT_CODE):
+                    errors = {"base": "unknown_event_code"}
+                else:
+                    event_code = device_data[CONF_EVENT_CODE]
+                    self._selected_device_event_code = event_code
+                    self._selected_device = self._config_entry.data[CONF_DEVICES][
+                        event_code
+                    ]
+                    self._selected_device_object = get_rfx_object(
+                        self._selected_device_event_code
+                    )
+                    return await self.async_step_set_device_options()
+            if CONF_REMOVE_DEVICE in user_input:
+                device_data = self._get_device_data(user_input[CONF_REMOVE_DEVICE])
+                if not device_data.get(CONF_EVENT_CODE):
+                    errors = {"base": "unknown_event_code"}
+                else:
+                    event_code = device_data[CONF_EVENT_CODE]
+                    device_id = device_data[CONF_DEVICE_ID]
+                    self.hass.helpers.dispatcher.async_dispatcher_send(
+                        f"{DOMAIN}_{CONF_REMOVE_DEVICE}_{device_id}"
+                    )
+                    device_registry = await async_get_registry(self.hass)
+                    device_registry.async_remove_device(user_input[CONF_REMOVE_DEVICE])
+                    devices = {event_code: None}
+                    self.update_config_data(
+                        global_options=self._global_options, devices=devices
+                    )
+
+                    return self.async_create_entry(title="", data={})
```

Comment:
I'm thinking maybe we should abort instead to avoid updating the entry twice.

Reply:
Hmm...can you verify? I think I'm hitting a core issue.

In `config_entries.py` in `async_finish_flow` it's doing:

```
if result["data"] is not None:
  self.hass.config_entries.async_update_entry(entry, options=result["data"])
```

However `_create_abort_data` in `data_entry_flow.py` does not write "data" so this throws a KeyError

```
@callback
def _create_abort_data(
    flow_id: str,
    handler: str,
    reason: str,
    description_placeholders: Optional[Dict] = None,
) -> Dict[str, Any]:
    """Return the definition of an external step for the user to take."""
    return {
        "type": RESULT_TYPE_ABORT,
        "flow_id": flow_id,
        "handler": handler,
        "reason": reason,
        "description_placeholders": description_placeholders,
    }
```

Reply:
Aside from this, we cannot use abort because it is visible in the UI.

![image](https://user-images.githubusercontent.com/31007358/89037461-cda81800-d33e-11ea-9e08-ada2cfa16cb3.png)

Reply:
We can set the abort reason string to show.

@balloob should we fix the options flow manager async_finish_flow to handle abort, or should we do something else?

Normally we save data in the options flow, but in this case the idea is to update the config entry data in the options flow instead. That's unconventional and why we're hitting this issue.

Reply:
We can set a string, but it would still show "Aborted" instead of "Finished" right? It really needs to show the latter.

The problem is not really updating instead of saving. We have two issues:
1. We must reload integration after options flow has finished, which internally calls `async_update_entry`. The listener that would normally be used for that however cannot be used because we also trigger `async_update_entry` to keep administration of new devices that are detected from rfx events.
2. Through option flow the `options` key is set and not the `data` key. As we still support import, it's not that straightforward to get device data in the `options` key as it's impossible to set this key from either import or config flow.

I had thought about finding a solution for the second, but it still would not give me a full solution so it seemed not worth the effort at this time.

Reply:
i really think we should find a way to get data into the options fields instead. Can't we load it into .data on import. On first time we start an option flow, we copy all data over to the options field instead, once there is data in the options field, we use that as a source for data?

Reply:
We can do that, but what does it bring to do that?

We still have to call update entry and reload manually, because we can't use the listener.

To me it seems like adding a lot of code which only adds complexity.

More generally speaking, I've seen PR's where others get creative as well to do these kind of copy actions, shouldn't it be considered then to be able to set options from import and config flow?

Reply:
We would not need to update the entry, since that's done when the options flow is finished. We would just need to reload the entry.

I think it's more in line with our current design to use options.data as a source of truth for the options. That's what we normally do right now.

Here's the related architecture issue. We shouldn't wait for that.
https://github.com/home-assistant/architecture/issues/377

Reply:
Ok, I thought we needed to update the entry before we call `self.hass.async_create_task(async_update_options(self.hass, self._config_entry))` but if we don't need to and can let the options flow update the entry it's better to do so.

We would need to wait for #38463. Without it there is no clean way to get the data in the options key in the first place.

Do note that the 0.113 release doesn't store options in the `data` key, but the current dev branch does. If we want to avoid making migration functions for ConfigEntries we may need some traction on that topic.

Reply:
If #38463 gets in, it seems better to make a separate PR to move data to `options` to get the ConfigEntry sorted.